rfcs/0042-omnilock/0042-omnilock.md

```diff
@@ -0,0 +1,558 @@
+```
+version (1 byte)
+current supply (16 bytes, little endian number)
+max supply (16 bytes, little endian number)
+sUDT script hash (32 bytes, sUDT type script hash)
+... (variable length, other data)
+```
+
+Currently, the version is 0. Only the current supply field can be updated during transactions. The script iterates all
+input and output cells, accumulating input amounts and output amounts identified by sUDT script hash. Then the script
+verifies:
+
+```
+<issued amount> = <output amount> - <input amount>
+<output current supply> = <issued amount> + <input current supply>
+```
+and
+```
+<output current supply> <= <max supply>
+```
+
+All the modes mentioned above can co-exist in Omnilock args in memory layout.
+
+## Omnilock Witness
+
+When unlocking an Omnilock, the corresponding witness must be a proper `WitnessArgs` data structure in molecule format. In
+the lock field of the `WitnessArgs`, an `OmniLockWitnessLock` structure must be present as follows:
+```
+import xudt_rce;
+
+array Auth[byte; 21];
+
+table Identity {
+    identity: Auth,
+    proofs: SmtProofEntryVec,
+}
+option IdentityOpt (Identity);
+
+// the data structure used in lock field of witness
+table OmniLockWitnessLock {
+    signature: BytesOpt,
+    omni_identity: IdentityOpt,
+    preimage: BytesOpt,
+}
+```
+
+When `omni_identity` is present, it will be validated whether the provided auth in `omni_identity` is present in RC
+AdminList Cell associated with the current lock script via SMT validation rules. In this case, the auth included in
+`omni_identity` will be used in further validation.
+
+If `omni_identity` is missing, the auth included in lock script args will then be used in further validation.
+
+Once the processing above is successfully done and the auth to be used is confirmed, the flag in the designated auth
+will be checked for the succeeding operations:
+
+* When the auth flag is 0x0, a signature must be present in `OmniLockWitnessLock`. We will use the signature for secp256k1
+  recoverable signature verification. The recovered public key hash using the blake160 algorithm must match the current
+  auth content.
+
+* When the auth flag is 0xFC, we will check against the current transaction, and there must be an input cell, whose lock
+  script matches the auth content when hashed via blake160.
+
+
+When `signature` is present, the signature can be used to unlock the cell in anyone-can-pay mode.
+
+When `preimage` is present, the auth flag can be 0xFD or 0xFE.
+
+* When auth flag is 0xFD, the `exec` method is used. The preimage's memory layout will be as follows:
+```
+exec code hash (32 bytes)
+exec hash type (1 byte)
+place (1 byte)
+bounds (8 bytes)
+pubkey blake160 hash (20 bytes)
+```
+
+The `place` and `bounds` are passed in directly to [exec](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0034-vm-syscalls-2/0034-vm-syscalls-2.md#exec):
+```C
+int ckb_exec(size_t index, size_t source, size_t place, size_t bounds, int argc, char* argv[]);
```

Comment:
Since the source can only be cell deps, why do we need to set the place parameter? It is only legal if it is 0, suggest to remove the `place (1 byte)` from the preimage

Reply:
This change requires the implementation to be updated. But the implementation has been deployed.
If we have further version of omnilock, I think it's OK to accept this change.